ogc_ws/src/despatcher/src/gnd_despatcher.py

```diff
@@ -45,87 +101,104 @@ def __init__(self):
-            if self.link_select == 0:
+            prefixes = ["i", self._is_air, self._id]
+            msg.data = headers.attach_headers(prefixes, [rospy.get_rostime().secs], data.data)
+            link = self._aircrafts[data.id].link_status()
+            if link == TELE:
                 self.pub_to_telegram.publish(msg)
-            elif self.link_select == 1:
+            elif link == SMS:
                 self.pub_to_sms.publish(msg)
-            else:
+            elif link == SBD:
                 self.pub_to_sbd.publish(msg)
-            self.pub_to_rqt_ondemand.publish("Command sent: " + data.data)
+            else:
+                rospy.logerr("Error: Invalid Link")
+                self.pub_to_rqt_ondemand.publish(\
+                    headers.attach_headers(dummy_prefixes, [rospy.get_rostime().secs], "Invalid Link: " + data.data))
+                return
+            self.pub_to_rqt_ondemand.publish(\
+                headers.attach_headers(dummy_prefixes, [rospy.get_rostime().secs], "Command sent: " + data.data))
 
     ###########################################
     # Handle Air-to-Ground (A2G) messages
     ###########################################
-
-    def _is_new_msg(self, timestamp):
-        '''Return true is incoming msg is a new msg'''
-        if timestamp < self._prev_transmit_time:
-            return False
-        else:
-            self._prev_transmit_time = timestamp
-            return True
     
     def check_incoming_msgs(self, data):
         '''Check for incoming A2G messages from ogc/from_sms, from_sbd or from_telegram topics'''
         try:
             # Handle msg prefixes
-            entries = data.data.split()
-            sender_timestamp = int(entries[-1])
-            sender_msgtype = str(entries[0])
-            sender_id = int(entries[2])
-            if not self._is_new_msg(sender_timestamp):
+            msgtype, devicetype, sysid, timestamp, entries \
+                = headers.split_headers(data.data)
+            if not self._new_msg_chk.is_new_msg(timestamp, sysid):
                 # Check if it is new msg
                 return
-            if regular.is_regular(sender_msgtype, len(entries)):
-                self.rqt_recovery_log(entries, sender_id)
+            if regular.is_regular(msgtype, len(entries)):
                 # Check if it is regular payload
                 msg = regular.convert_to_rosmsg(entries)
-                
+                self.rqt_recovery_log(entries, sysid)
                 self.pub_to_rqt_regular.publish(msg)
             else:
-                if sender_msgtype == 's':
+                if msgtype == 's':
                     # Check if it is statustext
                     self.pub_to_statustext.publish(data.data)
-                elif sender_msgtype == 'm':
+                elif msgtype == 'm':
                     # Check if it is mission update message
-                    mission_files = data.data.split()[3:-1]
```

Comment:
still need to define mission files

Reply:
Agreed, but we should use the entries variable created earlier instead; entries is the incoming message split into a list of words. That way we avoid defining multiple variables for the same list.